File: C++/week1/MaxPairwIseProductFast.cpp
#include<bits/stdc++.h>

using namespace std;

long long MaxPairwiseProductFast(const vector<int>& numbers) {
    int n = numbers.size();

    int max_index_1 = -1;
    for(int i = 0; i < n; i++)
        if((max_index_1 == -1) || (numbers[i] > numbers[max_index_1]))
            max_index_1 = i;

    int max_index_2 = -1;
    for(int i = 0; i < n; i++)
        if((i != max_index_1) && ((max_index_2 == -1) || (numbers[i] > numbers[max_index_2])))
            max_index_2 = i;

    return (long long) numbers[max_index_1] * numbers[max_index_2];
}

int main() {
    int n;
    cin>>n;
    vector<int> numbers(n);
    for(int i = 0; i < n; i++) {
        cin>>numbers[i];
    }
    long long result = MaxPairwiseProductFast(numbers);
    cout<<result<<endl;
    return 0;
}

File: Java/src/com/algorithms/week1/MaxPairwiseProductFast.java
package com.algorithms.week1;

import java.util.Scanner;

public class MaxPairwiseProductFast {

	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int n = scan.nextInt();
		int[] arr = new int[n];
		for(int i = 0;i < n; i++) {
			arr[i] = scan.nextInt();
		}
		long result = maxPairwiseProductFast(arr);
		System.out.println(result);
		scan.close();
	}
	
	public static long maxPairwiseProductFast(int[] arr) {
		int n = arr.length;
		
		int maxIndex1 = -1;
		for(int i = 0; i < n; i++) 
			if((maxIndex1 == -1) || (arr[i] > arr[maxIndex1]))
				maxIndex1 = i;
		
		int maxIndex2 = -1;
		for(int i=0; i< n; i++)
			if((i != maxIndex1) && ((maxIndex2 == -1) || (arr[i] > arr[maxIndex2])))
				maxIndex2 = i;
		
		return (long) arr[maxIndex1] * arr[maxIndex2];
	}

}
